Cover fullsize fixes

ronso0 opened the pull request:
These are some UX improvements for the full-size cover art windows:

* full-size cover in Track Properties dialog: changes are applied only when clicking Apply in Track Properties
* always update visible full-size covers when parent cover label changed
* in decks / library: keep size and position of full-size cover if a new track (with cover image) is loaded / selected


This is #11109 rebased onto 2.3 with 09cca5ed32a23c5627626693bdc36a2077a4eba3 and deb95b39f3218055312672bcf212da9f1f3a3403 from main.

I'll take care of merging to main, and I can also take care of rebasing #4851.

Commits are as small as possible, so review should be easy.

___
**Notes for testing:**

Full-size cover in decks (cover & spinny cover), library and Track Properties dialog:
* ~~image is updated in place (fit into the existing window) when a new track is loaded or selected via Next/Previous, or when a new image is assigned via the cover menu (from full-size, WCoverArt and WCoverArtLabel widgets or from any track menu)~~ post-ponde
* (new/fix) cover widget next to tracks table is updated if that track's cover is changed in Track Properties
* when full-size cover  is updated it is not 'activated', i.e. does not get keyboard focus, that is still on the previously focused widget
* full-size is closed if new track has no cover or if the cover is cleared via any related cover menu 

In Track Properties dialog:
* cleared/reset/new cover selected via cover menu (full-size or label in dialog) is applied only when pressing Apply
* full-size cover: if cover is cleared/reset/changed via the full-size menu, both that and the cover in Track Properties should be updated (track cover, e.g. in decks) should not change until Apply is clicked

In Tag Fetcher dialog:
* full-size is closed when a new MusicBrainz result is selected

**TODO**
- [ ] remove debug commit

Closes #10815
Closes #10816

ronso0 committed:
WCoverArtLabel: connect to customContextMenuRequested

ronso0 committed:
WTrackMenu: remove unused include wcoverartlabel.h

ronso0 committed:
cove art widgets: get rid of '!= nullptr'

ronso0 committed:
DlgCoverArtFullSize: split up slotLoadTrack() to use components separately

ronso0 committed:
DEBUG output for cover art widgets/dialogs

ronso0 committed:
cover art widgets: always update menu with new Coverinfo #10815

ronso0 commented:
@fatihemreyildiz Do you mind taking a look? Can you confirm all full-size use cases work as expected? Did I overlook one case?

fatihemreyildiz commented:
Hey @ronso0 thanks for all these fixes. 

I will test and confirm all use cases work as expected as soon as I can 👍

fatihemreyildiz commented:
I've did some quick tests according to "notes for testing" part. Thanks for all the work, cover arts operations feels more solid now. 

There is only one think that I noticed, when I right click on Spinny to populate full size, I hit the assertion on [dlgcoverartfullsize.cpp:263](https://github.com/mixxxdj/mixxx/blob/45da703b65d9b09171e05fc1758ff9b4289a88b5/src/library/dlgcoverartfullsize.cpp#L263)

ronso0 commented:
Thank you!
> There is only one think that I noticed, when I right click on Spinny to populate full size, I hit the assertion on [dlgcoverartfullsize.cpp:263](https://github.com/mixxxdj/mixxx/blob/45da703b65d9b09171e05fc1758ff9b4289a88b5/src/library/dlgcoverartfullsize.cpp#L263)

Fixed.

ronso0 commented:
Hmm I'll have to think about that.
Maybe the irritation stems from from the bright background fill? we could use the frame color for that.
Another option would be to adjust the window so that the new cover is fit into a square determined by the longer side of the current cover. For example if you start with a square cover and load a track with a portrait cover it would reduce the width. Switching to a landscape cover would make it grow horizontally to the width of the initial square.
Anyhow,, I think the resize is too much magic to be included in this Bugfix PR,so I'll write a feature request.

fatihemreyildiz commented:
As I said that was just an idea. This BugFix PR is already good and no need to make it bigger IMHO. Both of the ideas, filling background from the frame or resizing horizontally sounds nice. Leaving a request sounds fair 👍

ronso0 commented:
On second thought, the visible background is actually a regression of "update in place, don't resize dialog".
I'll see what I can do.

ronso0 commented:
> Another option would be to adjust the window so that the new cover is fit into a square determined by the longer side of the current cover. 

Alright, that was easier than expected. I added it to 676432c96c54dc137b683c299f17f0c4014fd84b
![vokoscreen-2022-12-09_01-25-14](https://user-images.githubusercontent.com/5934199/206594717-14871e1c-7ab4-41f8-8642-ae5644acbe5f.gif)

@fatihemreyildiz Please try.

fatihemreyildiz commented:
Just tried it and it is working well, there is no blank on the edges of the full sizes. 

Thank you 🐙

daschuer commented:
I have tested this a bit and it seems to work as intended, but I am not sure which use case it implements. 

For me, the main use case of the big cover is, to verify it in full screen. It is nice, that I can now do it on mass basis without double clicking the small cover every time. 

However it dos not work, because the next cover dos no longer show up in full size, it adopts the previous cover size. 
The other think is that the window closes and does not pop up again when one track has no cover. 

What is your use case?

ronso0 commented:
The main use case of "update in place, don't resize" for me is having the cover static when going through tracks in the  Track Properties dialog and in the library. Doing the same for the decks was just for consistency (i.e. I just didn't add a special handling for that):
If I want to see (show) the deck cover for some time I'd scale it down and move it out of the way, though when loading another track with cover art, the current behaviour after track change (maximized cover, centered on screen) just doesn't make sense to me because the cover would obscure my view again when I actually want to prepare the track.

If you want to view it 1:1 (or as big as possible) close and reopen it. In Track Properties just click the small cover twice.

> The other think is that the window closes and does not pop up again when one track has no cover.

Yeah, probably it should stay open with the default cover for consistency.

If you think this is too much of change for the stable version I can boil this down to the bugfixes and open another PR with the UX changes.

daschuer commented:
OK I understand. This feature does not feel consistent both use cases. 
Case 1: 
* Open full size cover = show full size
* Select next = show hat the size of the previous <- this feels inconsistent, expected: full size

Case 2: 
* Open full size cover = show full size
* Manually resize window to make the most of your screen space.
* Close view 
* Re-open full size cover = show full size <- this feels inconsistent, expected: size and position is kept. 

Not sure what is the best to fix this. Options:  
1. Always use full size 
2. Always use previous size and position + additional action to show full size (e.g. double click / magnifier mouse like in Firefox)

For now I prefer 1. Can we cover your use case differently? Maybe with a big track properties dialog?

ronso0 commented:
Maybe we can distinguish deck, library and Properties:
* deck: remember size and position (requires fixing
> // If you zoom in so the window is larger then the desktop, close the
// window and reopen, the window will not be resized correctly
// by slotCoverFound. Setting a small fixed size when hiding fixes this.
* Properties: always maximize

Or maybe the cover always remains maximized as long as the user didn't zoom that specific cover view?

Double-click to maximize would also be nice. Magnifier (single left-click) wouldn't work IMO since the cover window should still be easily operatable with mouse and touchscreen, quick open/close with single click.

I'll add some WIP commits so we can test the proposed behaviour.

daschuer commented:
@ronso0 What is the status here? Is this mandatory for Mixxx 2.3?

ronso0 commented:
Yes, this resolves quite a few issues so it should go into 2.3

I'll remove the debatable 'resize in place commit' commit 8d227011abd71e9861015ca955d972b1ebf52d71 , then this is ready for a final review.

daschuer commented:
@ronso0 Is this a 2.3.5 candidate?

ronso0 commented:
Could be, but I can't promise I'll finish it soon.
I could reduce it to the first bugfix commits, or it just has to wait for 2.3.6 or 2.4-beta

ronso0 committed:
WCoverArtLabel: update DlgFullSize with new cover if visible

ronso0 commented:
> I'll remove the debatable 'resize in place commit' commit https://github.com/mixxxdj/mixxx/commit/8d227011abd71e9861015ca955d972b1ebf52d71 , then this is ready for a final review.

Done.
I tested the UX again and fixed a remaining minor bug e1a762adc7e1ba70caabb384c42d58c302d897cd

All good now. @daschuer please have another look.
I updated the **Notes for testing** https://github.com/mixxxdj/mixxx/pull/11112#issue-1477513424

I'll test how well this merges into 2.4

ronso0 committed:
DlgCoverArtFullSize: don't set cover via menu if child of DlgTrackInfo

which already handles the track menu signals

ronso0 committed:
DlgCoverArtFullSize: show pending (not saved) cover of WCoverArtLabel if child of DlgTrackInfo

helps verifying the correct (high-res) cover was selected in Track Porperties

ronso0 committed:
WSpinnybase: remove obsolete member

ronso0 committed:
cover art widgets: use parented_ptr

ronso0 committed:
cover art widgets: add documentation

ronso0 committed:
cover art widgets: parent -> pParent

ronso0 commented:
> I'll test how well this merges into 2.4

not that well unfortunately..
Since I'm not aware of any code review I rebased onto 2.4. Should be easy to review.

ronso0 commented:
pre-commit is run on _all_ files?

daschuer commented:
> pre-commit is run on all files?

That is strange, I will have a look.

daschuer commented:
Strange. Your original commit fails in https://github.com/mixxxdj/mixxx/pull/11749 as well but after my dummy commit it works. I guess it is an obscure GitHub caching issue we can ignore for now.

daschuer commented:
I did another test with the track properties dialog.
Now the full cover is updated at 100 %. Which should cover the main use case to check the cover. 

However, if a track has no cover, the cover of the previous track is shown in a small version. This needs to be fixed.
I think showing the default/empty cover instead is reasonable in that case to not require to open the window again. 

The usability can improved, because in case of big covers, the cover may hide the preferences window. 
Not sure what the new solution is in that case. Showing the cover behind the dialog makes it IMHO worse. 

I think not re-centering the window for each new cover is probably good enough. If you want to check a single cover, the center position on top is good. If you want to check multiple covers, I think the users can effort to move the two windows to suitable positions and then start to step through their collections. In that state the window position should be kept.

ronso0 commented:
> The usability can improved, because in case of big covers, the cover may hide the preferences window.
Not sure what the new solution is in that case. Showing the cover behind the dialog makes it IMHO worse.

Yes, that's a difficult situation. If we want to cover the fullsize-check scenario (new cover is shown 1:1) and aim to keep the dialog visible (at least the Prev/Next buttons), it's incredible overkill to check the cover/dialog overlap and try to restore that.
Instead, let's assume a reasonable size range (1200*1200 max) and try to show the new cover centered on the center of the previous cover, while rsiking that it may be pushed off screen a bit.

ronso0 commented:
> However, if a track has no cover, the cover of the previous track is shown in a small version. This needs to be fixed.
I think showing the default/empty cover instead is reasonable in that case to not require to open the window again.

Jup, that seems feasable.

daschuer commented:
> it's incredible overkill to check the cover/dialog overlap and try to restore that.

I had not such a solution in mind. It would be already helpful to draw the next cover with the same top-left position as he previous one was drawn. This way users can arrange the two windows, in a way they want and the position is kept if the cover sizes are similar.

ronso0 commented:
> It would be already helpful to draw the next cover with the same top-left position as he previous one was drawn. 

Okay, so top-left is just a guess and users would need to adapt to that.
Not so far from the original update-in-place solution. If we assume users have track Properties and cover side by side for checking the size/quality, we could check in which screen half the image center is and then make the opposite edge stick. Seems feasible, just the in-between empty cover needs to be stretched to keep the side preference and not screw up auto-positioning.

Do you think this still a 2.4 candidate?

daschuer commented:
Yes, since it includes bug fixed. 
But don't stress us. A otherwise it can go to 2.4.1

daschuer commented:
If the image is bigger than the screen resoluton, the proposed check will go wrong. So we need something more sophisticated. For instance something that compares the distance to the screen borders and makes the shortest sticky. 

A simple solution that sticks one if the cover art window corners might be sufficient for the first attempt.

Swiftb0y commented on .swp:
whats this?

ronso0 replied:
Some Vim editor file. I have no clue if that is true, don't use vim or how it ended up in a commit.
I'll remove the debug commit anyway.
